### ISSUE TITLE
Coordinator updates: Implement `update_app`

### ISSUE THREAD

**Issue by jost-s** [holochain/holochain#4570]
Design: [`docs/design/coordinator_upgrades.md`](https://github.com/holochain/holochain/blob/develop/docs/design/coordinator_upgrades.md).

Add `update_app(installed_app_id, app_bundle) -> AppInfo`: an app-level operation that reconciles an installed app to the state described by a new bundle. It replaces `update_coordinators`, which is per-DNA and knows nothing about apps, roles or capability grants. Depends on #4404.

## Acceptance criteria

- `update_app` takes an installed app id and an app bundle, and treats the bundle as the complete desired state rather than a delta.
- Every role is validated before anything is applied. If any role fails validation, the installation is untouched.
- A role whose effective DNA identity has moved fails with `MigrationNotSupported` naming that role. A coordinator change is never applied to such a role.
- The apply step is atomic. A failure part-way leaves the app on its pre-upgrade state, including tearing down a cell instantiated for a newly added role.
- Coordinators are reconciled by name: present in both is a replace, bundle-only is an install, installed-only is a removal.
- Role-level fields are reconciled too. A field that cannot change in place fails the role rather than leaving the installation off the desired state.
- A role added by the bundle is provisioned per its provisioning strategy, and a role absent from it is removed.
- `AppInfo` reflects the new manifest after a successful update.
- `AdminRequest::UpdateCoordinators` and its payload, the client method, the sweettest helper, and the supporting bundle types are removed.
- Tests are written that cover the criteria above by providing happy/unhappy app updates.

## Suggested breakdown

- Happy path only, with an upgrade of an app that is enabled and doesn't have clone cells or membrane proofs
- Deal with edge cases
- Remove update coordinators

## Answered questions from below

- **Install-time role settings.** `InstallAppPayload` carries `network_seed` and `roles_settings`; `update_app` as designed carries neither. Any app installed with a modifier override would recompute a different `DnaHash` from the bundle and be told it needs a migration on what is really a coordinator-only change. The modifiers are not allowed to change.
- **Membrane proofs for a newly added role.** A new role is provisioned as it would be at install, but memproofs arrive in the install payload or through the deferred flow, and an update has neither. `update_app` accepts membrane proofs, and routes roles that need them through the deferred state.
- **App lifecycle.** Does an app has to be disabled to be updated? No, because calls that must go to a specific version can later be made by coordinator hash. What happens to an app waiting on deferred memproofs? The update is denied. Are zome calls are held while the update applies? They shouldn't need to be, calls go to the old coordinators until the atomic switch over.

**Comment by matthme** [comment]
> check that the DNA hash matches, because this operation may not update integrity zomes

What is "the DNA hash" referring to here? I think we need to get clear on terminology here. If we're talking about the hash of the cell including modifiers (which is what "dna hash" is referring to in the holochain code base as it stands) then I think this requirement is not correct because the modifiers in the manifest may differ from the modifiers of the installed app and an update would still be okay.

So I think we will also require a hash of a role without modifiers to become a thing (which we may want to still call "dna hash" and rename what is currently named "dna hash", i.e. the thing including modifiers, to "cell hash" or something like that).
